README


OpenMicroblogger -- Standards-based Microblog
Copyright (C) 2009 Brian Hendrickson

This application is free software; you can redistribute it and/or
modify it under the terms of the MIT License.

This application is distributed in the hope that it will be useful,
but without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

Author
 Brian Hendrickson - http://brianhendrickson.com
 
Project Manager
 Johannes Schirge - http://tiquer.me


Version 0.1, 1-Sep-2008
 initial release

Version 0.2, 22-Oct-2008
 sms, app store
 
Version 0.3, 10-July-2009
 admin, themes, intranet, search


REQUIREMENTS

   - PHP > 4.2
   - MySQL or PostgreSQL


INSTALLATION

/////////////////////
// 7 steps to install
/////////////////////

1. obtain OpenMicroblogger http://dbscript.net/omb.zip

2. unzip the files on your computer

3. copy the files to your Web server folder

4. open your Web browser and run
      yourdomain.com/install.php

5. follow the installation instructions

6. if installation succeeded, open your Web browser and browse to the installation folder

7. set in app/config/config.yml your 

      site_title: Your Site Title
      site_subtitle: Your Site Subtitle
      site_description: Your Site Description

 [[ that is it! ]]


/////////////////////////////////
// Facebook, Twitter, Zeep Mobile
/////////////////////////////////

Facebook:
  
  1. http://www.facebook.com/developers/apps.php
  2. click "Edit Settings"
  3. click "Connect"
your connect URL is
    http://yoursite.com/?email_login   [or]
    http://yoursite.com/email_login (pretty URLs)
    
Twitter:
  
  1. http://twitter.com/oauth_clients
your callback URL is
    http://yoursite.com/?oauth_login   [or]
    http://yoursite.com/oauth_login (pretty URLs)
  
Zeep Mobile:
  
  1. https://secure.zeepmobile.com/account/create_app
your callback URL is
    http://yoursite.com/?mobile_event   [or]
    http://yoursite.com/mobile_event (pretty URLs)
your website URL is
    http://yoursite.com/?mobile_settings   [or]
    http://yoursite.com/mobile_settings (pretty URLs)


/////////////////////////////
// use pretty URLs (optional)
/////////////////////////////

To enable pretty URLs:

1) edit config.php (or config/config.php if you used Web installer)

2) change the 2 lines at the bottom of config.php

    // global $pretty_url_base;
    // $pretty_url_base = "http://openmicroblogger.com";

  [[ change the 2 lines like this ]]

    global $pretty_url_base;
    $pretty_url_base = "http://yoursite.com";
    
3) copy the file /resource/prettyurls/.htaccess to the top level folder

4) change these 2 lines in .htaccess

  RewriteCond %{HTTP_HOST} ^www.openmicroblogger.com$ [NC]
  RewriteRule ^(.*)$ http://openmicroblogger.com/$1 [R=301,L]
  
  [[ change the 2 lines like this ]]

  RewriteCond %{HTTP_HOST} ^www.yoursite.com$ [NC]
  RewriteRule ^(.*)$ http://yoursite.com/$1 [R=301,L]
  
5) if you need to use [www.]yoursite.com or [blog.]yoursite.com
  
  [[ use this 1 line instead of 2 in your.htaccess ]]
  
  RewriteRule ^(.*)$ http://blah.yoursite.com/$1 [R=301,L]

Notes:
   "AllowEncodedSlashes On" is required in the VirtualHost (Apache > 2.0.46)
      todo: add a test for this problem XXX


///////////////////////////////////////
// (optional) choose a file upload mode
///////////////////////////////////////

1) upload FILES to /uploads (do not create BLOBs) (default)

      set permissions on the /cache and /uploads folders
      
      next, edit app/config/config.yml
   
   look for
   
      collection_cache:
      - posts:
        duration : 0
        location : cache    // and change 'cache' to 'uploads'

2) upload as SQL BLOBs
      
      maximum upload size is 5MB or so, not that great for big MP3s, but maximizes privacy
      
      you must set the /cache folder permissions so the Flash uploader can put files there

3) upload as SQL BLOBs (WITH CACHING)

      first set up your /cache folder as described in option 2
      
      then, edit app/config/config.yml

   look for

      collection_cache:
      - posts:
        duration : 0        // and change this 0 to the number of seconds to cache
        location : cache

4) upload to Amazon S3 (do not create BLOBs)

      set permissions on the /cache folder
      
      then, edit app/config/config.yml
   
   look for
   
      collection_cache:
      - posts:
        duration : 0
        location : cache    // and change 'cache' to 'aws'

   look for

      awsAccessKey:         // and put your AWS login info here
      awsSecretKey: 
      awsBucket: 


UPDATING

/////////////////////////////
// use the /wp-content folder
/////////////////////////////

If you want to customize OpenMicroblogger with your own theme and plugins:
  
1) copy /app/config/config.yml to /wp-content/config

2) duplicate a theme in /wp-content/themes

3) optionally, add plugins to /wp-content/plugins

    note: plugins in this folder override same-name plugins in /app/plugins


///////////////////////////
// get the latest build
///////////////////////////

you can grab the almost-nightly build from http://dbscript.net/omb.zip



//////////////////

OTHER DOCUMENTATION

1. changing the maximum number of tags shown in the sidebar

  The /app/omb/views/categories/_block.js file can be edited to change this setting.
  At the top of this file the $category_count variable is set to 10 by default.

2. login timeout can be set by changing "cookielife" in config.yml

3. creating a Contact Page/Impressum/About page: as an administrator you will have an app called "Pages"
   use this app to add new pages, links to these will appear in the footer of your P2-themed site

4. to change the allowed file-types for uploads, in config.yml is the "upload_types" list

5. configuring twitter-follow feature: in the site Admin panel is a new tab called 'Sources'
   choose from the full list of your site's OAuth-authenticated twitter users
   set "friends_timeline" on the users you want to import automatically
   here are the steps to set up the automatic twitter feed:
     1. see instructions above to set your "callback url" at Twitter and put key/secret in config.yml
     2. authenticate to the site as the Twitter user whose friends you want to import automatically
     3. set up a cron job to run db/library/remy-tweed/cron.php -- this example runs once a minute:
             0-59 * * * * /usr/bin/php /home/webuser/public_html/db/library/remy-tweed/cron.php 2>&1




changelog.txt
################################
# Openmicroblogger - Changelog #
################################

################################
#           Cutline            #
# * Changes                    #
# [#00] Ticket ID              #
################################

################################
# Current Release: 0.3b95      #
################################



This has changed in Release:

#0.3b95

* added: Realtime Wiki
* added: RSS Cloud
* added: Realtime comet function
* added: Linkify
* added: Tinymce
* added: personal RSS Feed
* added: flickr app
* updated: P2 Theme update
* fixed: Zeep Mobile
* fixed: Nickname already in Use bug


#0.3b94

* added: Default language feature


#0.3b93

* added: Language import feature


#0.3b92

* fixed: URL Shortener setup
* updated: Login Page


#0.3b91

* fixed: URL Shortener setup


#0.3b90

* added: URL Shortener


#0.3b89

* fixed: rply_yourls


#0.3b88

* added: rply_yourls support


#0.3b87

* added: User BG


#0.3b86

* install.php rollback


#0.3b85

* added: Memcache support


#0.3b84

* fixed: Streams


#0.3b83

* fixed: Tables
* fixed: Twitter app


#0.3b82

* fixed: subdomain issues


#0.3b81

* Readme updated
* Facebook Connect [#86]
* 6x6 followers grid on Frontpage P2-Theme
* http-auth fixes
* changing the maximum number of tags shown in the sidebar [#99]
* fixed: standard tag-it textstring is added to tag-list [#97]
* fixed: links to remote subscriptions on following page [#96]
* Added a stream delete button [#64]
* Only admin is able to enter admin-area [#59]
* fixed: remote-subscription-issue [#60]
* fixed: if you copy P2-Theme the profile picture was shown twice [#36]
* Set Photo-size in admin menu [#94]
* possibility to choose which filetypes can be uploaded [#58]
* more image-types can be used for profile-picture [#23]
* upgrade/fixed cookie auth [#69]
* possibility to add Contact box / impressum [#27]
* Some smaller fixes


#0.3b80

* rss submits current sitename/-description
* Fixed: Installer checks now if uploads-folder is writable
* Some OpenID-Changes
* Security fix: encrypt/decrypt twitter/identi.ca passwords [#83]
* some multilanguage improvements / import-function for exported lang-files [#61] 
* Fixed: In threaded-view post aren't shown twice anymore [#79]
* Fixed: register with password-mode [#77]
* Fixed: Zeep-Mobile settings are shown if available [#85]
* Fixed: Admin is now able to edit/delete every message [#9]
* Changed standard BG-Color in P2-Theme
* Added avatar-borders in P2-Theme
* Styled Reply-Page
* Added message-search feature [#87]
* P2-Theme: Hiding until now not existing features on sidebar/navigation
* Changing timezone feature: server (admin) and user-specific (settings) [#26]
* If user changes nickname @oldnickname redirects to @newnickname [#62]
* Fixed: @admin*** nicknames now redirecting to profile instead of admin-menu [#89]
* Added Facebook-Connect



#0.3b79

* P2-Theme: Sidebar: exchanged followers/following
* P2-Theme: Minor style.css fix
* Visual fix: new translation
* Fixed: Can't save avatars


#0.3b78

* possibility to disable/enable fileupload in admin menu
* possibility to set fileupload max-filesize in admin menu
* fixed: Error viewing first time the personal section
* added true/false instead of 1/0 in admin menu settings
* Cleaned up reply/permalink page


#0.3b77

* fixed remote-subscription-links
* fixed threaded-view-issues
* fixed: Tags are now saved as new categories
* fixed some language-creator issues
* added some features to language-creator


#0.3b76

* P2-Theme: Design-Update
* P2-Theme: following added to sidebar
* W3C-xhtml-conform fixes
* minor fixes
* more files added to multilanguage-support
* updated readme
* added this changelog so you can see what has changed since last release


#0.3b75

* Added Photo-upload in P2-Theme
* P2-Theme: Sidebar categories css update


#0.3b74

* omb now looks for @nickname at twitter.com if @nickname is not available
* fixed sub/unsub button
* Added twitter-like navigation to P2-Theme if Tag-Function is activated
* Replaced sidebar table/list style with div-style
* More inline-css fixes
* overhaul code: xhtml-conform <br />
* removed unused files


#0.3b73

* multilanguage fixes
* removed a lot of inline-css abd added to css-file


#0.3b72

* P2-Theme: Added sidebar-bg-color in css-file
* P2-Theme: Added sidebar-multilanguage-support
* P2-Theme: Header update
* P2-Theme: Textcounter style added
* German language: white site after posting fix 
* Design-Update of Reply/Permalink-Page
* fixed subscriptions list
* fix pagination, personal tab 


#0.3b71

* fixed language-selection


#0.3b70

* Security-Fix: Edit admin-menu-settings if not logged in as admin
* Added post2ff
* A lot of SVN changes
* A lot of language-updates and fixes
* fixed and enhanced installer
* P2-Theme: small sidebar design-update
* P2-Theme: added country, locality, website to sidebar and profile-edit
* P2-Theme: added sidebar greeting-message if not logged in
* Settings - Profile headline redesign


#0.3b69

* Language fixes


#0.3b68

* fixed remote subscriptions


#0.3b67

* Every user can choose a language by himself


#0.3b66

* Default language added


#0.3b65

* New-Function added: User translations


#0.3b64

* Set language on webinstall


#0.3b63

* New-Function added: Web-Installer


#0.3b62

* fixed installation error


#0.3b61

* followers/following links


#0.3b60

- skiped


#0.3b59

* P2-Theme: Navigation fixes
* P2-Theme: follow-button added


#0.3b58

* Multi-language support enhanced
* Minor fixes


#0.3b57

* Multi-language support added
* German localization


#0.3b56

- restored ##0.3b54 release


#0.3b55

- defective release


#0.3b54

* post fix


#0.3b53

* hide tags
* Copyright update
* Config-File added



#0.3b52 / #0.3b51

* openid overhaul


#0.3b50

* @replies, openid fix, registration fix, subscribers, subscriptions


#0.3b49 - 0.1

- not listed


db/boot.php
<?php
   
  /** 
   * dbscript -- restful openid framework
   * @version 0.6.0 -- 22-October-2008
   * @author Brian Hendrickson <brian@dbscript.net>
   * @link http://dbscript.net/
   * @copyright Copyright 2009 Brian Hendrickson
   * @package dbscript
   * @license http://www.opensource.org/licenses/mit-license.php MIT License
   */
   
  /**
   
   dbscript -- restful openid framework
   Copyright (C) 2008 Brian Hendrickson
   
   This library is free software; you can redistribute it and/or
   modify it under the terms of the MIT License.
   
   This library is distributed in the hope that it will be useful,
   but without any warranty; without even the implied warranty of
   merchantability or fitness for a particular purpose.
   
   Author
     Brian Hendrickson - http://brianhendrickson.com
   
   Version 0.1, 16-Nov-2006
     initial release at pdxphp meeting
   
   Version 0.1.1, 15-Feb-2007
     added gary court's content-negotiation library
   
   Version 0.1.2, 19-Feb-2007
     changed object instantiation routine
   
   Version 0.1.3, 23-Feb-2007
     fixed call-time pass-by-reference error in View
   
   Version 0.2.0, 19-Mar-2007
     added openid authentication to security plugin
   
   Version 0.3.0, 10-Jun-2007
     models for Group, Membership, Identity
   
   Version 0.5.0, 12-August-2008
     new templates: vcard, hcard, ics, rdf, json, atom
   
   Version 0.6.0, 22-October-2008
     apps, openappstore
   
   */
   
$version = '0.6.0';

  /**
   * directory paths
   */

global $views,$app,$config,$env,$exec_time,$version,$response;
global $variants,$request,$loader,$db,$logic;

  /**
   * load config
   */

if (file_exists('config/config.php'))
  require('config/config.php');
else
  require('config.php');

  /**
   * memcached
   */

if (MEMCACHED) {
  $perma = parse_url( $_SERVER['REQUEST_URI'] );
  $_PERMA = explode( "/", $perma['path'] );
  @array_shift( $_PERMA );
  if ( isset($_PERMA[0]) && $_PERMA[0] != basename($_SERVER['PHP_SELF']) ){
    require_once 'db/library/pca/pca.class.php';
    $cache = PCA::get_best_backend();
    if ( $cache->exists( $_SERVER['REQUEST_URI'] )) {
  		header( 'Location: '.$cache->get( $_SERVER['REQUEST_URI'] ), TRUE, 301 );
  		exit;
  	}
  }
  require_once 'db/library/pca/pca.class.php';
  $cache = PCA::get_best_backend();
  $_SERVER['FULL_URL'] = 'http://';
  if ( $_SERVER['SERVER_PORT']!='80' ) {
    $port = ':' . $_SERVER['SERVER_PORT'];
  }
  if ( isset( $_SERVER['REQUEST_URI'] ) ) {
    $script = $_SERVER['REQUEST_URI'];
  } else {
    $script = $_SERVER['PHP_SELF'];
    if ( $_SERVER['QUERY_STRING']>' ' ) {
      $script .= '?'.$_SERVER['QUERY_STRING'];
    }
  }
  if ( isset( $_SERVER['HTTP_HOST'] ) ) {
    $_SERVER['FULL_URL'] .= $_SERVER['HTTP_HOST'] . $port . $script;
  } else {
    $_SERVER['FULL_URL'] .= $_SERVER['SERVER_NAME'] . $port . $script;
  }
  global $pretty_url_base;
  if (isset($pretty_url_base) && !empty($pretty_url_base)) {
    if (!empty($_SERVER['QUERY_STRING']))
      $url = $pretty_url_base.'/?'.$_SERVER['QUERY_STRING'];
    else
      $url = $pretty_url_base.'/';
  } else {
    $url = $_SERVER['FULL_URL'];
  }
  if($cache->exists($url) && $cache->exists($url.'type')){
    header( 'Content-Type: '.$cache->get($url.'type') );
    header( "Content-Disposition: inline" );
    echo $cache->get($url);
    exit;
  }
}

  // set path to db directory
if (is_dir('db'))
  $app = 'db' . DIRECTORY_SEPARATOR;
elseif (is_dir('site' . DIRECTORY_SEPARATOR . 'db' . DIRECTORY_SEPARATOR))
  $app = 'site' . DIRECTORY_SEPARATOR . 'db' . DIRECTORY_SEPARATOR;
else
  trigger_error( 'path to dbscript not found', E_USER_ERROR );

$GLOBALS['PATH'] = array();
$GLOBALS['PATH']['app'] = $app;
$GLOBALS['PATH']['library'] = $app . 'library' . DIRECTORY_SEPARATOR;
$GLOBALS['PATH']['controllers'] = $app . 'controllers' . DIRECTORY_SEPARATOR;
$GLOBALS['PATH']['models'] = $app . 'models' . DIRECTORY_SEPARATOR;
$GLOBALS['PATH']['plugins'] = $app . 'plugins' . DIRECTORY_SEPARATOR;
$GLOBALS['PATH']['dbscript'] = $GLOBALS['PATH']['library'] . 'dbscript' . DIRECTORY_SEPARATOR;

  /**
   * load dbscript minimal functions & classes
   */

foreach( array(
    '_functions',
    'bootloader',
    'mapper',
    'route',
    'genericiterator',
    'collection',
    'view',
    'cookie'
  ) as $module ) {

  include $GLOBALS['PATH']['dbscript'] . $module . '.php';
  
}

  // load HTTP_Negotiate by Gary Court
lib_include( 'http_negotiate' );

  // load Cake's inflector
lib_include( 'inflector' );


error_reporting( E_ALL & ~E_NOTICE & ~E_WARNING );
$dbscript_error_handler = set_error_handler( 'dbscript_error' );


  /**
   * cross-platform magic-quotes init
   */

  // turn off magic quotes
@set_magic_quotes_runtime(0);

  // if get_magic_quotes_gpc, strip quotes or slashes
if ( get_magic_quotes_gpc() ) {
  if ( @ini_get( 'magic_quotes_sybase' )=='1' ) {
    $_GET = magic_quotes_stripquotes($_GET);
    $_POST = magic_quotes_stripquotes($_POST);
    $_COOKIE = magic_quotes_stripquotes($_COOKIE);
    $_REQUEST = magic_quotes_stripquotes($_REQUEST);
  } else {
    $_GET = magic_quotes_stripslashes($_GET);
    $_POST = magic_quotes_stripslashes($_POST);
    $_COOKIE = magic_quotes_stripslashes($_COOKIE);
    $_REQUEST = magic_quotes_stripslashes($_REQUEST);
  }
}


  /**
   * routes
   */

  // create a request mapper object to regex-match the URI to a Route
$request = new Mapper();

  // add a new Route
$request->connect(
  
  // route pattern
  'static/:staticresource',
  
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+' )
  )

);
// load static-file-cache and debug aspects
//include $GLOBALS['PATH']['plugins'] . 'renderer.php';

// this doesn't do anything because the aspect-filter was deleted XXX
//$request->routematch();


$request->connect(
  'admin',
  array(
    'action'=>'index',
    'resource'=>'admin'
  )
);


  /**
   * read the configuration file
   */

include $GLOBALS['PATH']['library'] . 'yaml.php';

$loader = new Horde_Yaml();

if ( file_exists( $app . 'config.yml' ) ) {
  extract($loader->load(file_get_contents($app.'config.yml')));
  extract( $$env['enable_db'] );
} else {
  $env = array('app_folder'=>'app');
}


// if app folder exists, re-config for the app

if (is_dir( $env['app_folder'] )) {
  $app = $env['app_folder'] . DIRECTORY_SEPARATOR;
  $appdir = $app;
  if ( file_exists( $app . 'config' . DIRECTORY_SEPARATOR . 'config.yml' ) ) {
    extract($loader->load(file_get_contents($app . 'config' . DIRECTORY_SEPARATOR .'config.yml')));
    extract( $$env['enable_db'] );
    if (isset($env['boot']))
      $appdir = $app.$env['boot'].DIRECTORY_SEPARATOR;
    else
      $appdir = $app.'omb'.DIRECTORY_SEPARATOR;
    $GLOBALS['PATH']['app'] = $app;
    $app = $appdir;
    $GLOBALS['PATH']['controllers'] = $appdir . 'controllers' . DIRECTORY_SEPARATOR;
    $GLOBALS['PATH']['models'] = $appdir . 'models' . DIRECTORY_SEPARATOR;
  }
  if (is_dir( $appdir . 'plugins' . DIRECTORY_SEPARATOR ))
    $GLOBALS['PATH']['plugins'] = $appdir . 'plugins' . DIRECTORY_SEPARATOR;
}




// debug mode

if ($env['debug_enabled']) {
  ini_set('display_errors','1');
  ini_set('display_startup_errors','1');
  error_reporting (E_ALL & ~E_NOTICE );
  global $exec_time;
  $exec_time = microtime_float();
}


// set up wp-config folder

$content_config = 'wp-content'.DIRECTORY_SEPARATOR.'config'.DIRECTORY_SEPARATOR.'config.yml';

if (file_exists( $content_config )) {
  
  // if    exists /wp-content/config/config.yml
  // then  use /wp-content
  
  // theme MUST be present in /wp-content/themes
  
  // plugins from /wp-content/plugins will take precedent
  // todo wp-plugins not supported yet in this plugins folder
  
  extract( $loader->load( file_get_contents( $content_config )));
  extract( $$env['enable_db'] );
  
}


$wp_theme = "wp-content".DIRECTORY_SEPARATOR."themes".DIRECTORY_SEPARATOR.$env['theme'];

if ((file_exists($wp_theme))) {
  $GLOBALS['PATH']['content_plugins'] = 'wp-content'.DIRECTORY_SEPARATOR.'plugins'.DIRECTORY_SEPARATOR;
  $GLOBALS['PATH']['themes'] = "wp-content".DIRECTORY_SEPARATOR."themes".DIRECTORY_SEPARATOR;
} else {
  $GLOBALS['PATH']['themes'] = $request->template_path . 'wp-themes' . DIRECTORY_SEPARATOR;
}

// set up the content-negotiation template paths

if ( is_dir( $app . $env['view_folder'] ) )
  $request->set_template_path( $app . $env['view_folder'].DIRECTORY_SEPARATOR );
else
  $request->set_template_path( $env['view_folder'].DIRECTORY_SEPARATOR );
  
if ( is_dir( $app . $env['layout_folder'] ) )
  $request->set_layout_path( $app . $env['layout_folder'].DIRECTORY_SEPARATOR );
else
  $request->set_layout_path( $env['layout_folder'].DIRECTORY_SEPARATOR );


  /**
   * connect to the database with settings from config.yml
   */

  // load dbscript database support classes
db_include( array(
  'database',
  'model',
  'record',
  'recordset',
  'resultiterator',
  $adapter
));

if (DB_NAME)
  $database = DB_NAME;

if (DB_USER)
  $username = DB_USER;

if (DB_PASSWORD)
  $password = DB_PASSWORD;

if (DB_HOST)
  $host = DB_HOST;


  // init the Database ($db) object and connect to the database
$db = new $adapter(
  $host,
  $database,
  $username,
  $password
);


/**
 * boot the loads
 */

$loader = new BootLoader();

$loader->start();

if ( $db->just_get_objects() )
  return;


/**
 * connect pre-plugin routes
 */

// doesn't work XXX
//$request->connect( 'migrate' );


/**
 * set up wp theme and plugin paths
 */

$wp_theme = "wp-content".DIRECTORY_SEPARATOR."themes".DIRECTORY_SEPARATOR.$env['theme'];

if ((file_exists($wp_theme))) {
  $GLOBALS['PATH']['content_plugins'] = 'wp-content'.DIRECTORY_SEPARATOR.'plugins'.DIRECTORY_SEPARATOR;
  $GLOBALS['PATH']['themes'] = "wp-content".DIRECTORY_SEPARATOR."themes".DIRECTORY_SEPARATOR;
} else {
  $GLOBALS['PATH']['themes'] = $env['themepath'.$env['theme']].DIRECTORY_SEPARATOR;
}


/**
 * OMB MU setup
 */

$params = array_merge($_GET,$_POST);
$stream = false;
list($subdomain, $rest) = explode('.', $_SERVER['SERVER_NAME'], 2);
// XXX subdomain upgrade
if ($pretty_url_base && !mu_url() && !('http://'.$subdomain.".".$rest == $pretty_url_base)) {
  $request->base = 'http://'.$subdomain.".".$rest;
  $request->domain = $subdomain.".".$rest;
  $pretty_url_base = $request->base;
  $stream = $subdomain;
// XXX subdomain upgrade
} elseif (mu_url()) {
  $pattern='/(\?)?twitter\/([a-z]+)(\/?)/';
  if ( 1 <= preg_match_all( $pattern, $request->uri, $found )) {
	  if ($pretty_url_base && environment('subdomains')){
		  $trail = "/";
		  $pattern2='/(\?)?twitter\/([a-z]+)(\/?)(\/.+)/';
		  if ( 1 <= preg_match_all( $pattern2, $request->uri, $found2 ))
        $trail = $found2[4][0];
      redirect_to('http://'.$found[2][0].".".$request->domain.$trail);
	  }
    $uri = $request->uri;
    $tags[] = $found;
    // XXX subdomain upgrade
    $repl = 'twitter/'.$tags[0][2][0].$tags[0][3][0];
    $request->uri = str_replace($repl,'',$uri);
    $request->prefix = $repl;
    $request->setup();
    $trail = '';
    if (empty($tags[0][3][0]))
      $trail = "/";
    $request->base = substr($uri,0,strpos($uri,$tags[0][0][0])+(strlen($repl)+1)).$trail;
  }
  $stream = $tags[0][2][0];
} elseif (isset($params['username']) && isset($params['password']) && !isset($_FILES['media'])) {
  $sql = "SELECT nickname,profile_id FROM shorteners WHERE nickname LIKE '".$db->escape_string($params['username'])."'";
  $sql .= " AND password LIKE '".$db->escape_string($params['password'])."'";
  $result = $db->get_result( $sql );
  if ( $db->num_rows($result) == 1 ) {
    if (!($pretty_url_base && !mu_url() && !('http://'.$subdomain.".".$rest == $pretty_url_base))) {
      $request->base = 'http://'.$subdomain.".".$request->domain;
      $request->domain = $subdomain.".".$request->domain;
      $pretty_url_base = $request->base;
    }
    $stream = $db->result_value( $result, 0, "nickname" );
    global $db,$request;
    global $person_id;
    global $api_methods,$api_method_perms;
    if (array_key_exists($request->action,$api_method_perms)) {
      $arr = $api_method_perms[$request->action];
      if ($db->models[$arr['table']]->can($arr['perm']))
        return;
    }
    $Identity =& $db->get_table( 'identities' );
    $Person =& $db->get_table( 'people' );
    $i = $Identity->find($db->result_value( $result, 0, "profile_id" ));
    $p = $Person->find( $i->person_id );
    if (!(isset( $p->id ) && $p->id > 0)) {
      header('HTTP/1.1 401 Unauthorized');
      echo 'BAD LOGIN';
      exit;
    }
    $person_id = $p->id;
  }
}
if ($stream) {
  if (!$db->table_exists('blogs')) {
    $Blog =& $db->model('Blog');
    $Blog->save();
  }
  $sql = "SELECT prefix FROM blogs WHERE nickname LIKE '".$db->escape_string($stream)."'";
  $result = $db->get_result( $sql );
  if ( $db->num_rows($result) == 1 ) {
    global $prefix;
    $prefix = $db->result_value( $result, 0, "prefix" )."_";
    $db->prefix = $prefix;
  }
}


/**
 * load saved config
 */

$Setting =& $db->model('Setting');
$Setting->find_by(array(
  'eq'    => 'like',
  'name'  => 'config%'
));
while ($s = $Setting->MoveNext()) {
  $set = split('\.',$s->name);
  if (is_array($set) && $set[0] == 'config') {
    if ($set[1] == 'env') {
      $env[$set[2]] = $s->value;
    } elseif ($set[1] == 'perms') {
      $tab =& $db->models[$set[2]];
      if ($tab)
        $tab->permission_mask( $set[3],$s->value,$set[4] );
    }
  }
}


/**
 * overrides from config.php
 */

if (isset($env['max_upload_mb']))
  $db->max_upload_megabytes($env['max_upload_mb']);

if (INTRANET)
  $env['authentication'] = 'password';

//if (UPLOADS)
//  $env['collection_cache']['posts']['location'] = UPLOADS;

//if (UPLOADS)
//  $env['collection_cache']['identities']['location'] = UPLOADS;

// PHP5 only set server timezone
if (function_exists(timezone_abbreviations_list) && environment('timezone'))
  if (setting('timezone'))
    set_tz_by_offset(setting('timezone'));
  else
    set_tz_by_offset(environment('timezone'));


/**
 * load virtual API methods
 */

global $api_methods,$api_method_perms;
$api_methods = array();
$api_method_perms = array();
$Method =& $db->model('Method');
$Method->find_by(array(
  'eq'        => 'like',
  'function'  => 'api_%'
));
while ($m = $Method->MoveNext()) {
  $api_method_perms[$m->function] = array('table'=>$m->resource,'perm'=>$m->permission);
  $api_methods[$m->function] = $m->code;
  $patterns = explode( '/', $m->route );
  $requirements = array();
  foreach ( $patterns as $pos => $str ) {
    if ( substr( $str, 0, 1 ) == ':' ) {
		  $requirements[] = '[A-Za-z0-9_.]+';
    }
  }
	$routesetup = array(
	  'action'=>$m->function,
	  'resource'=>$m->resource
	);
	if (count($requirements) > 0)
		$routesetup['requirements'] = $requirements;
  $request->connect(
    $m->route,
    $routesetup
  );
  if ($m->omb)
    before_filter( 'authenticate_with_omb', $m->function );
  if ($m->http)
    before_filter( 'authenticate_with_http', $m->function );
  if ($m->oauth)
    before_filter( 'authenticate_with_oauth', $m->function );
}


/**
 * load plugins
 */

if ( isset( $env ))
  while ( list( $key, $plugin ) = each( $env['plugins'] ) )
    load_plugin( $plugin );


/**
 * connect more Routes to the Mapper
 */

$request->connect(
  ':resource/:id/email/:ident',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[0-9]+', '[A-Za-z0-9]+' )
  )
);

$request->connect(
  ':resource/page/:page/:action',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[0-9]+','[A-Za-z0-9_.]+' )
  )
);

$request->connect(
  ':resource/page/:page',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[0-9]+' )
  )
);

$request->connect(
  ':resource/:id/:action',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[0-9]+', '[A-Za-z0-9_.]+' )
  )
);

$request->connect(
  ':resource/:id',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[0-9]+' )
  )
);

$request->connect(
  ':resource/:action',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[A-Za-z0-9_.]+' )
  )
);

$request->connect(
  ':resource/:id/:action/partial',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[0-9]+', '[A-Za-z0-9_.]+' )
  )
);

$request->connect(
  ':resource/:action/partial',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+', '[A-Za-z0-9_.]+' )
  )
);

$request->connect(
  ':resource',
  array(
    'requirements' => array ( '[A-Za-z0-9_.]+' )
  )
);

$request->connect( '', array( 'resource'=>$env['goes'], 'action'=>'get' ) );

  // (for debugging) print the current request variables and die
#aspect_join_functions( 'routematch', 'catch_params' );

$request->routematch();

//print_r($request->activeroute); echo '<br /><br />'; print_r($request->params); exit;

/**
 * attach functions to aspect crosscuts
 */

  // load data model if these model methods are triggered
before_filter( 'load_model', 'delete_from_post' );
before_filter( 'load_model', 'insert_from_post' );
before_filter( 'load_model', 'update_from_post' );
before_filter( 'load_model', 'fields_from_request' );
before_filter( 'load_model', 'MoveFirst' );
before_filter( 'load_model', 'MoveNext' );
before_filter( 'load_model', 'base' );
before_filter( 'load_model', 'find' );

  // add a filter to persist submitted data on error
before_filter( 'session_error', 'handle_error' );

  // activate Taint Mode to validate each input
before_filter( 'regex_validate', 'save_record' );

  // read the Access List and verify action permissions
before_filter( 'model_security', $request->action );

  // if public resource, ping the search index server
after_filter( 'send_ping', 'insert_from_post' );
after_filter( 'send_ping', 'update_from_post' );

  // echo value after single-field Ajax PUT call
after_filter( 'ajax_put_field', 'update_from_post' );
after_filter( 'ajax_put_field', 'insert_from_post' );


// authenticate yourself without OpenID

//test_log_in();

function test_log_in() {
  $person_id = 1;
  set_cookie($person_id);
  $_SESSION['openid_complete'] = true;
}


/**
 * negotiate the best content-type for the client
 */

$response = new View();

render( 'action', $request->action );

